ding/torch_utils/network/tests/test_resnet.py

```diff
@@ -10,3 +13,93 @@ def test_resnet18():
     inputs = torch.randn(4, 3, 224, 224)
     outputs = model(inputs)
     assert outputs.shape == (4, 1000)
+
+
+stem_type = ['', 'deep', 'deep,tiered']
+replace_stem_pool = [True, False]
+avg_down = [True, False]
+block = [BasicBlock]
+layers = [[2, 2, 2, 2]]
+zero_init_last_bn = [True, False]
+output_stride = [8, 32]
+num_classes = [0, 1000]
+args = [
+    item for item in
+    product(*[stem_type, replace_stem_pool, avg_down, block, layers, zero_init_last_bn, output_stride, num_classes])
+]
+
+
+@pytest.mark.unittest
+@pytest.mark.parametrize(
+    'stem_type, replace_stem_pool, avg_down, block, layers, zero_init_last_bn, output_stride, num_classes', args
+)
+def test_resnet(stem_type, replace_stem_pool, avg_down, block, layers, zero_init_last_bn, output_stride, num_classes):
+    model = ResNet(
+        stem_type=stem_type,
+        replace_stem_pool=replace_stem_pool,
+        avg_down=avg_down,
+        block=block,
+        layers=layers,
+        output_stride=output_stride,
+        num_classes=num_classes,
+        drop_rate=0. if stem_type == 'deep' else 0.05
+    )
+    model.init_weights(zero_init_last_bn=zero_init_last_bn)
+    inputs = torch.randn(4, 3, 224, 224).requires_grad_(True)
+    outputs = model(inputs)
+    assert outputs.shape == (4, num_classes if num_classes > 0 else 512)
+    mse_loss = torch.nn.MSELoss()
+    target = torch.randn(outputs.shape)
+    loss = mse_loss(outputs, target)
+    assert inputs.grad is None
+    loss.backward()
+    assert isinstance(inputs.grad, torch.Tensor)
```

Comment:
I think it is strange to examine the gradient of input. Why not check the gradient of the model?

Reply:
Good point!